platform/include/retry_utils.h

```diff
@@ -31,6 +31,143 @@
 /* Standard include. */
 #include <stdint.h>
 
+/**
+ * @page retryutils_page Retry Utilities
+ * @brief An abstraction of utilities for retrying with exponential back off and
+ * jitter.
+ *
+ * @section retryutils_overview Overview
+ * The retry utilities are a set of APIs that aid in retrying with exponential
+ * backoff and jitter. Exponential backoff with jitter is strongly recommended
+ * for retrying failed actions over the network with servers. Please see
+ * https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/ for
+ * more information about the benefits with AWS.
+ *
+ * Exponential backoff with jitter is typically used when retrying a failed
+ * connection to the server. In an environment with poor connectivity, a client
+ * can get disconnected at any time. A backoff strategy helps the client to
+ * conserve battery by not repeatedly attempting reconnections when it is
+ * unlikely to succeed.
+ *
+ * Before retrying the failed communication to the server there is a quiet period.
+ * In this quiet period, the task that is retrying must sleep for some random
+ * amount of seconds between 0 and the lesser of a base value and a predefined
+ * maximum. The base is doubled with each retry attempt until the maximum is
+ * reached.<br>
+ *
+ * > sleep_seconds = random_between( 0, min( 2<sup>attempts_count</sup> * base_seconds, maximum_seconds ) )
+ *
+ * @section retryutils_implementation Implementing Retry Utils
+ *
+ * The functions that must be implemented are:<br>
+ * - @ref RetryUtils_ParamsReset
+ * - @ref RetryUtils_BackoffAndSleep
+ *
+ * The functions are used as shown in the diagram below. This is the exponential
+ * backoff with jitter loop:
+ *
+ * @image html retry_utils_flow.png width=25%
+ *
+ * The following steps give guidance on implementing the Retry Utils. An example
+ * implementation of the Retry Utils for a POSIX platform can be found in file
+ * @ref retry_utils_posix.c.
+ *
+ * -# Implementing @ref RetryUtils_ParamsReset
+ * @snippet this define_retryutils_paramsreset
+ *<br>
+ * This function initializes @ref RetryUtilsParams_t. It is expected to set
+ * @ref RetryUtilsParams_t.attemptsDone to zero. It is also expected to set
+ * @ref RetryUtilsParams_t.nextJitterMax to @ref INITIAL_RETRY_BACKOFF_SECONDS
+ * plus some random amount of seconds, jitter. This jitter is a random number
+ * between 0 and @ref MAX_JITTER_VALUE_SECONDS. This function must be called
+ * before entering the exponential backoff with jitter loop using
+ * @ref RetryUtils_BackoffAndSleep.<br><br>
+ * Please follow the example below to implement your own @ref RetryUtils_ParamsReset.
+ * The lines with FIXME comments should be updated.
+ * @code{c}
+ * void RetryUtils_ParamsReset( RetryUtilsParams_t * pRetryParams )
+ * {
+ *     uint32_t jitter = 0;
+ *     myTimeType_t myTime; // FIXME: Replace with your system's time type.
```

Comment:
Could've just left it as `struct timespec tp;` since it's in `time.h` and I don't think it's POSIX specific

Reply:
Oh oops time.h is part of c90 sorry I should have look this up

Reply:
Actually I did further reserach and time,h is posix specific.

Reply:
For non-posix C89 and C90, there is no timespec structure.